src/rust/engine/fs/store/src/remote/reapi.rs

```diff
@@ -143,40 +140,79 @@ impl Provider {
 
     let mut client = self.byte_stream_client.as_ref().clone();
 
-    let chunk_size_bytes = self.chunk_size_bytes;
+    // we have to communicate the (first) error reading the underlying reader out of band
+    let error_occurred = Arc::new(parking_lot::Mutex::new(None));
+    let error_occurred_stream = error_occurred.clone();
 
-    let stream = futures::stream::unfold((0, false), move |(offset, has_sent_any)| {
-      if offset >= len && has_sent_any {
-        futures::future::ready(None)
-      } else {
-        let next_offset = min(offset + chunk_size_bytes, len);
-        let req = protos::gen::google::bytestream::WriteRequest {
+    let stream = if len == 0 {
```

Comment:
For next time, I think that using the `async_stream` crate and a loop is a much more obvious way to produce a stream like this (especially when a conditional gets involved like this): https://github.com/pantsbuild/pants/blob/a103211cd68a5fed7c7ac987a192e5e6c8d403de/src/rust/engine/process_execution/docker/src/docker.rs#L683-L714

Reply:
Ah, cool, I'll experiment with switching to that.